docs/content/en/docs/Design/simd.md

```diff
@@ -0,0 +1,175 @@
+This documentation will mostly focus on describing how the optimization is realized in HEIR (which differs somewhat from the original implementation) and how the optimization is intended to be used in an overall end-to-end compilation pipeline.
+
+
+## Representing FHE SIMD Operations
+Following the design principle of maintaining programs in standard MLIR dialects as long as possible  (cf. the design rationale behind the [Secret Dialect](https://heir.dev/docs/design/secret/)),
+HEIR uses the MLIR [`tensor` dialect](https://mlir.llvm.org/docs/Dialects/TensorOps/) and [ElementwiseMappable](https://mlir.llvm.org/docs/Traits/#elementwisemappable) operations from the MLIR [`arith` dialect](https://mlir.llvm.org/docs/Dialects/ArithOps/) to represent HE SIMD operations.
+
+We do introduce the HEIR-specific [`tensor_ext.rotate`](https://heir.dev/docs/dialects/tensorext/#tensor_extrotate-heirtensor_extrotateop) operation, which represents a cyclical left-rotation of a tensor.
+Note that, as the current SIMD vectorizer only supports one-dimensional tensors, the semantics of this operation on multi-dimensional tensors are not (currently) defined.
+
+For example, the common "rotate-and-reduce" pattern which results in each element containing the sum/product/etc of the original vector can be expressed as:
+```llvm
+%tensor = tensor.from_elements %i1, %i2, %i3, %i4, %i5, %i6, %i7, %i8 : tensor<8xi16>
+%0 = tensor_ext.rotate %tensor, %c4 : tensor<8xi16>, index
+%1 = arith.addi %tensor, %0 : tensor<8xi16>
+%2 = tensor_ext.rotate %1, %c2 : tensor<8xi16>, index
+%3 = arith.addi %1, %2 : tensor<8xi16>
+%4 = tensor_ext.rotate %3, %c1 : tensor<8xi16>, index
+%5 = arith.addi %3, %4 : tensor<8xi16>
+```
+The `%cN` and `%iN`, which are defined as `%cN = arith.constant N : index` and `%iN = arith.constant N : i16`, respectively, have been omitted for readability.
+
+## Intended Usage
+The `-heir-simd-vectorizer` pipeline transforms a program consisting of loops and index-based accesses into tensors (e.g., `tensor.extract` and `tensor.insert`) into one consisting of SIMD operations (including rotations) on entire tensors.
+While its implementation does not depend on any FHE-specific details or even the Secret dialect, this transformation is likely only useful when lowering a high-level program to an arithmetic-circuit-based FHE scheme (e.g., B/FV, BGV, or CKKS).
+The `-mlir-to-openfhe-bgv` pipeline demonstrates the intended flow: augmenting a high-level program with `secret` annotations, then applying the SIMD optimization (and any other high-level optimizations) before lowering to BGV operations and then exiting to OpenFHE.
+
+> **Warning**
+> The current SIMD vectorizer pipeline supports only one-dimensional tensors.
+> As a workaround, one could reshape all multi-dimensional tensors into one-dimensional tensors, but MLIR/HEIR currently do not provide a pass to automate this process.
```

Comment:
Useful warning! Next HEIR meeting I may pick brains at this. I started writing a packing pass that promotes arbitrary tensors into tensors with a desired power of two number of slots with padding/replication, but I'm a little stuck on multi-dimensional tensors (which is probably necessary for matrix multiplication). It seems to me slightly chicken-and-eggy to write a packing pass and then find optimal rotations which depend on the packing. But like you mention below, we will always be relying on heuristics, and so some kind of heuristic will probably be necessary (and maybe we allow for handwritten attributes or options that can indicate a better approach)